### PR TITLE
device: binding lookup should return null if the name is null

### PR DESCRIPTION
**Pull request by pabigot**
A null device name should map to a null device.

NB: #30951 was merged to test this, but it could never have passed.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

### CODE REVIEW

**Review by galak — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
It looks fine to me, now, thanks @pabigot.
@nashif could we get this fix in as soon as possible? The bug was found via #31481 and currently blocks me proceeding with it :)

**Comment by ioannisg** [comment]
> NB: #30951 was merged to test this, but it could never have passed.

#31481 should be able to catch these bugs, at least in Cortex-M. I am wondering though @andyross, I thought we have null-pointer dereferencing detection in x86, right? why this was not caught in QEMU?

**Comment by nashif** [comment]
how is this a hotfix? is this breaking the tree (master) right now? Maybe this should be high priority bug instead?

**Comment by pabigot** [comment]
As best I can tell the test was only verifying the API when invoked from userspace.  Correct it to also check calls not from userspace.

This may be wrong, but since the test only worked because the call was from userspace something needs to be done to check this.  Perhaps the test should be built in two different ways.

File: tests/kernel/device/src/main.c
Comment (nashif) [inline tests/kernel/device/src/main.c:473]: this should not needed, if userspace is not enabled, `ztest_user_unit..` behaves like  `ztest_unit...`

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/kernel/device/src/main.c
Comment (pabigot) [reply tests/kernel/device/src/main.c:473]: Yes, but the test is never run without userspace, is it?

Comment (nashif) [reply tests/kernel/device/src/main.c:473]: No it is not, but the fix should be in the testcase itself (which you did), not here.

Comment (ioannisg) [reply tests/kernel/device/src/main.c:473]: agreed

Comment (pabigot) [reply tests/kernel/device/src/main.c]: The test uses `CONFIG_TEST_USERSPACE=y` unconditionally.  AFAICT that implies `CONFIG_USERSPACE`.  The only reason the test passed at all was because it was conditional on `CONFIG_USERSPACE`, and the validation that made it work was only performed with `CONFIG_USERSPACE=y`.

So to catch the original error, it would have been necessary to both eliminate the source-code filtering of the test with `CONFIG_USERSPACE=n`, *and* to build the code with `CONFIG_USERSPACE=n`.

So from what I can tell we would need to modify this test (and potentially others) to build and run without userspace in order to catch future instances of this sort of problem.

**Comment by ioannisg** [comment]
> how is this a hotfix? is this breaking the tree (master) right now? Maybe this should be high priority bug instead?

No i don't think it does, that's why I asked for a quick resolution :)

**Comment by pabigot** [comment]
I marked it `hotfix` because at the time it wasn't clear how that test could have passed.  Now I know.

File: tests/kernel/device/src/main.c
Comment (nashif) [reply tests/kernel/device/src/main.c]: > The test uses `CONFIG_TEST_USERSPACE=y` unconditionally. AFAICT that implies `CONFIG_USERSPACE`.

all of this depends on what the platform can do, the same test can run without userspace on any platform.

**Review by dcpleung — APPROVED** [review]

File: kernel/device.c
Comment (ceolin) [inline kernel/device.c:82]: why not use `CHECKIF` ?

**Review by ceolin — CHANGES_REQUESTED** [review]
I'm not getting it, almost all Zephyr API's don't check for parameters because it's *assumed* that the application is trusted, we usually add an assert to the error be captured during the development. My question is why this API is different from, for instance, `k_thread_create` that does not check the pointers parameters as well ? Is this parameter coming from directly from an  untrusted source, e.g network. It does not seem to be case ...

Also,  the check in the verification function is wrong as well, AFAIK `z_user_string_copy` already handle NULL pointer.

Nevertheless, an assert check in the implementation function seems ok for me, have suggested using `CHECKIF` there.

**Comment by ioannisg** [comment]
> Also, the check in the verification function is wrong as well, AFAIK z_user_string_copy already handle NULL pointer.

@ceolin I am not sure I agree with this. z_user_string_copy seems to de-reference pointers, which may be NULL. I don't consider this as "handing". We should probably not de-reference pointers that are NULL, IMHO

**Comment by ceolin** [comment]
> > Also, the check in the verification function is wrong as well, AFAIK z_user_string_copy already handle NULL pointer.
> 
> @ceolin I am not sure I agree with this. z_user_string_copy seems to de-reference pointers, which may be NULL. I don't consider this as "handing". We should probably not de-reference pointers that are NULL, IMHO

The exception is gracefully handled. In this specific api when called from user thread it returns null, as expected. I didn't check other  arch but x86 though. But if I remember correctly, there are other syscalls relying on this behavior. If this is not happening we should fix there and not in device_get_binding IMHO

**Comment by ioannisg** [comment]
> > > Also, the check in the verification function is wrong as well, AFAIK z_user_string_copy already handle NULL pointer.
> > 
> > 
> > @ceolin I am not sure I agree with this. z_user_string_copy seems to de-reference pointers, which may be NULL. I don't consider this as "handing". We should probably not de-reference pointers that are NULL, IMHO
> 
> The exception is gracefully handled. In this specific api when called from user thread it returns null, as expected. I didn't check other arch but x86 though. But if I remember correctly, there are other syscalls relying on this behavior. If this is not happening we should fix there and not in device_get_binding IMHO

Oh, I see, what you mean, let me check what is the case for Cortex-M. Thanks :)

**Comment by ioannisg** [comment]
@ceolin I had a look and it seems Cortex-M also gracefully handles exceptions on z_user_string_copy, meaning that we do not seem to need the fix in the verification function. Won't harm, though, so I leave it up to the kernel maintainers.

**Comment by pabigot** [comment]
Switched to use `CHECKIF()`, and removed the check in the verify function since `z_user_string_copy()` is documented to return `-EFAULT` on memory access errors.

**Comment by pabigot** [comment]
With the merge of #32266 the problem is fixed, so now this only updates the test to be correct in a configuration that's not part of normal testing.

**Comment by ceolin** [comment]
I'm just seeing one commit now that is removing the guards, this is wrong, passing a null pointer to `device_get_binding` without memory protection will crash.

**Comment by pabigot** [comment]
> I'm just seeing one commit now that is removing the guards, this is wrong, passing a null pointer to `device_get_binding` without memory protection will crash.

No, it won't; see #32266.

**Comment by ceolin** [comment]
Ok, so that that NULL check got into the tree. What is your plan now, starting to add NULL checks in all APIs ?

**Comment by pabigot** [comment]
@ceolin is your changes-requested still relevant since the test does work correctly without usespace being enabled?

**Review by ceolin — APPROVED** [review]

**Comment by pabigot** [comment]
Rebased to fix the merge conflict.  (The blocking review's concerns were addressed over a month ago.)

There's been so much churn around this with the null pointer checks and trust zone, and the desire to not explicitly check for failures but instead rely on exceptions to detect problems, I honestly don't know if this is a correct fix anymore.  I note that `z_user_string_nlen` still documents its behavior as undefined in the case of error, which seems like a bad idea.

**Review by tbursztyka — APPROVED** [review]

**Comment by LeiW000** [comment]
@nashif , it looks that the PR can be merged now.

**Pull request closed without merging**
